packages/trie/src/db.ts

```diff
@@ -55,23 +106,79 @@ export class DB {
    * @param value The value to be stored
    */
   async put(key: Buffer, val: Buffer): Promise<void> {
+    const revertOps: BatchDBOp[] = []
+    // In CP mode check for an old value to be put
+    // on the revert ops stack
+    if (this.isCheckpoint) {
+      const oldValue = await this.get(key)
+      if (oldValue !== null) {
+        revertOps.push({
+          type: 'put',
+          key: key,
+          value: oldValue,
+        })
+      }
+    }
     await this._leveldb.put(key, val, ENCODING_OPTS)
+    // In CP mode add del to the revert ops stack
+    if (this.isCheckpoint) {
+      revertOps.push({
+        type: 'del',
+        key: key,
+      })
+      for (const revertOp of revertOps) {
+        this.addCPRevertOperation(revertOp)
+      }
+    }
   }
 
   /**
    * Removes a raw value in the underlying leveldb.
    * @param keys
    */
   async del(key: Buffer): Promise<void> {
+    const revertOps: BatchDBOp[] = []
+    // In CP mode check for an old value to be put
+    // on the revert ops stack
+    if (this.isCheckpoint) {
+      const oldValue = await this.get(key)
+      if (oldValue !== null) {
+        revertOps.push({
+          type: 'put',
+          key: key,
+          value: oldValue,
+        })
+      }
+    }
     await this._leveldb.del(key, ENCODING_OPTS)
+    if (this.isCheckpoint) {
+      for (const revertOp of revertOps) {
+        this.addCPRevertOperation(revertOp)
+      }
+    }
   }
 
   /**
    * Performs a batch operation on db.
    * @param opStack A stack of levelup operations
    */
   async batch(opStack: BatchDBOp[]): Promise<void> {
-    await this._leveldb.batch(opStack, ENCODING_OPTS)
+    if (this.isCheckpoint) {
+      for (const op of opStack) {
+        if (op.type === 'put') {
+          // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+          if (!op.value) {
```

Comment:
is this `!op.value` check for safety so in case it isn't present it doesn't run a del operation?

Reply:
Took this code over from `Trie.batch()` and thought it might be safer to leave (but not interrupt the `batch()` run, might be better on the `DB` level to just do the plain `level` operations "as-is", will remove.